File: AOC_3.py
import utils
import re

reg_mul = re.compile("mul\(([0-9]{1,3}),([0-9]{1,3})\)")
reg_cond = re.compile("don't\(\).*?(do\(\)|$)", re.DOTALL)


def solve_puzzle(filename):
    input_data = utils.read_file_as_chunk(filename)
    res_1 = [int(x[0]) * int(x[1]) for x in reg_mul.findall(input_data)]
    input_data_mod = reg_cond.sub("", input_data)
    res_2 = [int(x[0]) * int(x[1]) for x in reg_mul.findall(input_data_mod)]
    return sum(res_1), sum(res_2)


if __name__ == '__main__':
    p1, p2 = solve_puzzle("Test_3.txt")
    print(p1, p2)
    p1, p2 = solve_puzzle("Data_3.txt")
    print(p1, p2)

File: AOC_3.cpp
#include "utils.hpp"
#include <numeric>
#include <regex>

int get_muls(std::string& str){
    int ret = 0;
    std::string::const_iterator curr_start(str.cbegin());
    std::smatch match;
    const std::regex reg("mul\\(([0-9]{1,3}),([0-9]{1,3})\\)");
    while(std::regex_search(curr_start, str.cend(), match, reg)){
        ret += std::stoi(match[1].str()) * std::stoi(match[2].str());
        curr_start = match.suffix().first;
    }
    return ret;
}

std::pair<int, int> solve_puzzle(std::string filename){
    std::pair<int, int> res{0, 0};
    std::vector<std::string> lines = read_file_as_lines(filename);
    std::string text = std::accumulate(lines.begin(), lines.end(), std::string(""));
    res.first = get_muls(text);
    std::regex repl("don't\\(\\).*?(do\\(\\)|$)");
    text = std::regex_replace(text, repl, "");
    res.second = get_muls(text);

    return res;
}

int main(){
    std::pair<int, int> res = solve_puzzle("Test_3.txt");
    std::cout << "TEST res pt 1: " << res.first << " pt 2: " << res.second << "\n";
    
    res = profile_function(solve_puzzle, "Data_3.txt");
    std::cout << "Puzzle res pt 1: " << res.first << " pt 2: " << res.second << "\n";

    return 0;
}